### PR TITLE
faster reindex

### PR DESCRIPTION
**Pull request by yingjieMiao**
@ymatsuda

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/NormalizedURIRepo.scala [hunk 1]
```diff
@@ -68,6 +69,12 @@ extends DbRepo[NormalizedURI] with NormalizedURIRepo with ExternalIdColumnDbFunc
     (if (limit >= 0) q.take(limit) else q).list
   }
 
+  override def getCurrentSeqNum()(implicit session: RSession): SequenceNumber = {
+    val q = (for( r <- table ) yield r.seq)
+    val m = q.list.map{_.value}.reduce(_ max _)
+    SequenceNumber(m)
+  }
+
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/model/NormalizedURIRepo.scala:77]: any better way?

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala [hunk 2]
```diff
@@ -6,22 +6,29 @@ import com.keepit.search.article.ArticleIndexer
...
     resetSequenceNumberIfReindex()
-
     var total = 0
     var done = false
     while (!done && !closing) {
-      val uris = Await.result(shoeboxClient.getIndexableUris(sequenceNumber.value, fetchSize), 180 seconds)
+      val uris =  if (sequenceNumber.value >= catchUpSeqNumber.value) Await.result(shoeboxClient.getIndexableUris(sequenceNumber.value, fetchSize), 180 seconds)
+      else {
+        log.info(s"ShardedArticleIndexer in catch up mode: skip active uris until seq number passes ${catchUpSeqNumber.value}")
+        val uris = Await.result(shoeboxClient.getScrapedUris(sequenceNumber.value, fetchSize), 180 seconds)
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala]: Is it possible that getScrapedUri returns URIs that exceed catchUpSeqNumber?

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala [hunk 3]
```diff
@@ -6,22 +6,29 @@ import com.keepit.search.article.ArticleIndexer
...
+        if (uris.nonEmpty) uris else  { sequenceNumber = catchUpSeqNumber; return total }
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala:30]: Do we have to return here?

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala (hunk 2 above)
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala]: good catch!

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala (hunk 3 above)
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala:30]: not necessarily. Just indicating "catch up" is done. See a later test. I need to return explicitly here.

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala [hunk 4]
```diff
@@ -47,4 +58,11 @@ class ShardedArticleIndexer(
     if (uris.nonEmpty) sequenceNumber = uris.last.seq
     total
   }
+
+  override def reindex(): Unit = {
+    super.reindex()
+    val seqNum: SequenceNumber = SequenceNumber(Await.result(shoeboxClient.getHighestUriSeq(), 5 seconds))
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedArticleIndexer.scala:64]: This should be the min of committed seqNum all shards.

File: kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala [hunk 5]
```diff
@@ -119,8 +120,16 @@ abstract class Indexer[T](
     _sequenceNumber = n
   }
 
+  def catchUpSeqNumber_=(n: SequenceNumber) {_catchUpSeqNumber = n}
+
+  private[this] var _catchUpSeqNumber = SequenceNumber(commitData.getOrElse(Indexer.CommitData.sequenceNumber, "-1").toLong)
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/index/Indexer.scala:125]: max(
  commitData.getOrElse(Indexer.CommitData.sequenceNumber, "-1").toLong, 
  commitData.getOrElse(Indexer.CommitData.catchUpSequenceNumber, "-1").toLong
)

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala [hunk 6]
```diff
@@ -12,6 +12,7 @@ trait ShardedIndexer[K, T <: Indexer[_]] extends IndexManager[T] {
   @volatile protected var closing = false
 
   def commitSequenceNumber: SequenceNumber = SequenceNumber(indexShards.valuesIterator.map(indexer => indexer.commitSequenceNumber.value).min)
+  def commitCatchUpSeqNumber: SequenceNumber = SequenceNumber(indexShards.valuesIterator.map(indexer => indexer.commitCatchUpSeqNumber.value).min)
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala:15]: SequenceNumber(indexShards.valuesIterator.map{ indexer =>
  if (indexer.numDocs <= 0) Await.result(shoeboxClient.getHighestUriSeq(), 5 seconds)
  else indexer.commitCatchUpSeqNumber.value
}.min)

**Pull request merged**